File: leetcode/editor/cn/1201_丑数 III_test.go
package leetcode

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func Test_givenNormal_when1201_thenSuccess(t *testing.T) {
	assert.Equal(t, 4, nthUglyNumber(3, 2, 3, 5))
	assert.Equal(t, 6, nthUglyNumber(4, 2, 3, 4))
}

func Test_givenFailedCase1_when1201_thenSuccess(t *testing.T) {
	assert.Equal(t, 10, nthUglyNumber(5, 2, 11, 13))
}

func Test_givenFailedCase2_when1201_thenSuccess(t *testing.T) {
	assert.Equal(t, 1999999984, nthUglyNumber(1000000000, 2, 217983653, 336916467))
}

func Test_givenGcd_when1201_thenSuccess(t *testing.T) {
	assert.Equal(t, 7, gcd1201(98, 63))
	assert.Equal(t, 7, gcd120101(98, 63))
	assert.Equal(t, 52, gcd120101(260, 104))
	assert.Equal(t, 52, gcd1201(260, 104))
}

//leetcode submit region begin(Prohibit modification and deletion)
func nthUglyNumber(n int, a int, b int, c int) int {
	ans := 0
	// TODO 值域2分
	left, right := 0, nthUglyNumberMin1201(a, b, c)*n
	ab, ac, bc := lcm1201(a, b), lcm1201(a, c), lcm1201(b, c)
	abc := lcm1201(ab, c)
	for left <= right {
		mid := left + (right-left)>>1
		num := mid/a + mid/b + mid/c - mid/ab - mid/ac - mid/bc + mid/abc
		if num == n {
			if mid%a == 0 || mid%b == 0 || mid%c == 0 {
				ans = mid
				break
			}
			right = mid - 1
		} else if num > n {
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	return ans
}

func nthUglyNumberMin1201(args ...int) int {
	ans := args[0]
	for i := 0; i < len(args); i++ {
		if args[i] < ans {
			ans = args[i]
		}
	}
	return ans
}

// TODO  lcm gcd 都是针对正整数
/**
最小公倍数

least common multiple
*/
func lcm1201(n1, n2 int) int {
	return n1 * n2 / gcd1201(n1, n2)
}

/**
最大公约数

greatest common divisor

highest common factor
*/
// TODO 欧几里得 辗转相除法
func gcd1201(n1, n2 int) int {
	if n2 == 0 {
		return n1
	}
	return gcd1201(n2, n1%n2)
}

// TODO 算术九章 更相减损术
func gcd120101(n1, n2 int) int {
	if n1 < n2 {
		n1, n2 = n2, n1
	}
	subtract := n1 - n2
	if n2 == subtract {
		return subtract
	}
	return gcd120101(n2, subtract)
}

//leetcode submit region end(Prohibit modification and deletion)

File: leetcode/editor/cn/1201_丑数 III_test.py
import math
import unittest


class TestSolution(unittest.TestCase):

    def setUp(self) -> None:
        self.solution = Solution()

    def test_givenNormal_when_thenSuccess(self):
        left, right = 0, 6
        while left <= right:
            mid = left + ((right - left) >> 1)
            print(mid)
            if mid == 1:
                break
            elif mid > 1:
                right = mid - 1
            else:
                left = mid + 1
        ans = self.solution.nthUglyNumber(3, 2, 3, 5)
        self.assertEqual(4, ans)


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def nthUglyNumber(self, n: int, a: int, b: int, c: int) -> int:
        ans = 0
        left, right = 0, min(a, b, c) * n
        ab, ac, bc = math.lcm(a, b), math.lcm(a, c), math.lcm(b, c)
        abc = math.lcm(ab, c)
        while left <= right:
            # TODO >> 的优先级比 + 低
            mid = left + ((right - left) >> 1)
            ugly_num = mid // a + mid // b + mid // c
            ugly_num = ugly_num - mid // ab - mid // ac - mid // bc
            ugly_num += mid // abc
            if ugly_num == n:
                if mid % a == 0 or mid % b == 0 or mid % c == 0:
                    ans = mid
                    break
                else:
                    right = mid - 1
            elif ugly_num > n:
                right = mid - 1
            else:
                left = mid + 1
        return ans


# leetcode submit region end(Prohibit modification and deletion)


if __name__ == '__main__':
    unittest.main()
